RM-9250 Implement support for .NET DateOnly in ObjectBinding

a-ctor opened the pull request:
https://re-motion.atlassian.net/browse/RM-9250

MichaelKetting commented on Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocDateTimeValueTest.cs:
```diff
@@ -142,6 +142,17 @@ public void SetValueToNullableDateTimeNull ()
+      DateOnly dateOnly = new DateOnly(2006, 1, 1);
+      _bocDateTimeValue.IsDirty = false;
+      Assert.That(() => { ((IBusinessObjectBoundControl)_bocDateTimeValue).Value = dateOnly;},
+          Throws.InstanceOf<ArgumentException>()
+              .With.Message.EqualTo("Parameter 'value' has type 'System.DateOnly' when type 'System.Nullable`1[System.DateTime]' was expected. (Parameter 'value')"));
```
Please with to the dedicated ArgumentException assertion:
`Throws.ArgumentException.With.ArgumentExceptionMessageEqualTo(messsage, parametername)`

MichaelKetting commented on Remotion/ObjectBinding/UnitTests/BusinessObjectStringFormatterServiceTests/GetPropertyString_WithDateTimeProperty.cs:
```diff
@@ -71,6 +71,25 @@ public void Scalar_WithValu_AndDateProperty ()
       }
     }
 
+#if NET6_0_OR_GREATER
+    [Test]
+    public void Scalar_WithValue_AndDateOnlyProperty ()
+    {
+      _mockProperty.Setup(_ => _.Type).Returns(DateTimeType.Date).Verifiable();
+      _mockProperty.Setup(_ => _.IsList).Returns(false).Verifiable();
+      _mockBusinessObject.Setup(_ => _.GetProperty(_mockProperty.Object)).Returns(new DateOnly(2000, 6, 17)).Verifiable();
+
+      using (new CultureScope(new CultureInfo("de-de"), new CultureInfo("de-de")))
+      {
+        string actual = _stringFormatterService.GetPropertyString(_mockBusinessObject.Object, _mockProperty.Object, null);
+
+        _mockBusinessObject.Verify();
+        _mockProperty.Verify();
+        Assert.That(actual, Is.EqualTo("17.06.2000"));
+      }
+    }
+#endif
+
```
Turns out, we can revert this since the GetProperty should return DateTime, not DateOnly. Since we are mocking the property, there is no differernce between DateTimeProperty, DateProperty and DateOnlyProperty.

In the orignal specification, the assumption was we are using actual properties loaded via reflection from the domain in the test.

MichaelKetting commented on Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
+      var result = property.ConvertToNativePropertyType(new DateTime(2042, 4, 2, 13, 34, 20));
+
+      Assert.That(result, Is.InstanceOf<DateOnly>());
+      Assert.That(result!.As<DateOnly>(), Is.EqualTo(new DateOnly(2042, 4, 2)));
```
```suggestion
      Assert.That(result, Is.EqualTo(new DateOnly(2042, 4, 2)));
```

MichaelKetting commented on Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
+      var result = property.ConvertFromNativePropertyType(new DateOnly(2042, 4, 2));
+
+      Assert.That(result, Is.InstanceOf<DateTime>());
+      Assert.That(result!.As<DateTime>(), Is.EqualTo(new DateTime(2042, 4, 2)));
```
```suggestion
      Assert.That(result, Is.EqualTo(new DateTime(2042, 4, 2)));
```

MichaelKetting commented on Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:
```diff
@@ -16,6 +16,7 @@
 // 
 using System;
 using NUnit.Framework;
+using Remotion.Development.UnitTesting;
```
```suggestion
```

MichaelKetting commented on Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
+
+      Assert.That(property.Type, Is.EqualTo(DateTimeType.Date));
+    }
+
```
please add tests for NULL values (to and from native value)

MichaelKetting commented on Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs:
```diff
@@ -0,0 +1,35 @@
+
+    public override object? ConvertFromNativePropertyType (object? nativeValue)
+    {
+      if (nativeValue != null)
+        return ((DateOnly)nativeValue).ToDateTime(TimeOnly.MinValue);
+      return base.ConvertFromNativePropertyType(nativeValue);
```
the other proprties use
```suggestion
      var dateOnly = ArgumentUtility.CheckType<DateOnly?>("nativeValue", nativeValue);
      if (dateOnly != null)
        return (dateOnly.ToDateTime(TimeOnly.MinValue);
      return base.ConvertFromNativePropertyType(dateOnly);
```

MichaelKetting commented on Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs:
```diff
@@ -0,0 +1,35 @@
+
+    public override object? ConvertToNativePropertyType (object? publicValue)
+    {
+      if (publicValue != null)
+        return DateOnly.FromDateTime((DateTime)publicValue);
+      return base.ConvertToNativePropertyType(publicValue);
```
The other properties use
```suggestion
      var dateTime = ArgumentUtility.CheckType<DateTime?>("publicValue", publicValue);
      if (dateTime != null)
        return DateOnly.FromDateTime(dateTime );
      return base.ConvertToNativePropertyType(dateTime);
```

a-ctor commented:
@MichaelKetting I have addressed all the review comments but I haven't had the time to fix the screenshot tests. These need to be fixed for the build to succeed. Apart from that the PR is done
